packages/std/src/sections.rs

```diff
@@ -0,0 +1,64 @@
+/// A sections decoder for the special case of two elements
+#[allow(dead_code)] // used in Wasm and tests only
+pub fn decode_sections2(data: Vec<u8>) -> (Vec<u8>, Vec<u8>) {
+    let section2_len: usize = if data.len() >= 4 {
+        u32::from_be_bytes([
+            data[data.len() - 4],
+            data[data.len() - 3],
+            data[data.len() - 2],
+            data[data.len() - 1],
+        ]) as usize
+    } else {
+        panic!("Cannot read section2 length");
+    };
+
+    let section1_len_end = data.len() - 4 - section2_len;
+    let section1_len: usize = if section1_len_end >= 4 {
+        u32::from_be_bytes([
+            data[section1_len_end - 4],
+            data[section1_len_end - 3],
+            data[section1_len_end - 2],
+            data[section1_len_end - 1],
+        ]) as usize
+    } else {
+        panic!("Cannot read section1 length");
+    };
+
+    if data.len() != 4 + section1_len + 4 + section2_len {
+        panic!(
+            "Invalid data length: {}, {}, {}",
+            data.len(),
+            section1_len,
+            section2_len
+        );
+    }
+
+    let mut first = data;
+    let mut second = first.split_off(section1_len_end);
```

Comment:
I was thinking of an lower-level api that would let one easily create decodeN wrappers.

`split_tail(data: Vec<u8>) -> StdResult<(Vec<u8>, Vec<u8>)>` which would parse out the last section, allocate memory and truncate the input return (head, tail).

`one_element(data: Vec<u8>) -> StdResult<Vec<u8>>` after splitting off all the N args, this is called for the last (first) item. It parses out the length bytes, ensures they are correct, and then returns the truncated input without a realloc.

We would end up with something like:

```rust
pub fn decode_sections2(mut data: Vec<u8>) -> StdResult<(Vec<u8>, Vec<u8>)> {
  let (head, second) = split_tail(data)?;
  let first = one_element(head)?;
  Ok((first, second))
}

pub fn decode_sections3(mut data: Vec<u8>) -> StdResult<(Vec<u8>, Vec<u8>)> {
  let (head, third) = split_tail(data)?;
  let (head, second) = split_tail(head)?;
  let first = one_element(head)?;
  Ok((first, second, third))
}
```

Reply:
This would make sense if we intend 3, 4, 5 variants and avoids any bugs in writing those. Not sure if we will ever need those however

Reply:
This is pretty cool. Not because we need it not but because it explains the algorithm. Implemented. The caller code got even nicer:

```rust
pub fn decode_sections2(data: Vec<u8>) -> (Vec<u8>, Vec<u8>) {
    let (rest, second) = split_tail(data);
    let (_, first) = split_tail(rest);
    (first, second)
}
```